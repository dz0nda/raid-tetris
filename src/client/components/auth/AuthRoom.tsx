import React, { FC, useState } from 'react';
import { useForm } from '@mantine/form';
import {
  Button,
  Group,
  Paper,
  Stack,
  Switch,
  Text,
  TextInput,
  Title,
  createStyles,
  rem,
  useMantineTheme,
} from '@mantine/core';
import { IconCheck, IconX } from '@tabler/icons-react';

import { Column, TableSort } from '../table-sort/TableSort';
import { Room } from '@/client/interfaces/room.interface';
import { roomsListData } from '@/client/helpers/data';

// Styles
const useStyles = createStyles((theme) => ({
  card: {
    backgroundColor: theme.colorScheme === 'dark' ? theme.colors.dark[7] : theme.white,
  },
  item: {
    '& + &': {
      paddingTop: theme.spacing.sm,
      marginTop: theme.spacing.sm,
      borderTop: `${rem(1)} solid ${theme.colorScheme === 'dark' ? theme.colors.dark[4] : theme.colors.gray[2]}`,
    },
  },
  switch: {
    '& *': {
      cursor: 'pointer',
    },
  },
  title: {
    lineHeight: 1,
  },
}));

// Types
export interface AuthRoomFormValues {
  room: string;
  pass: string;
}

interface AuthRoomProps {
  username: string;
  onValidated: (values: AuthRoomFormValues) => void;
}

const columns: Column<Room>[] = [
  { key: (row) => row?.room || '', label: 'Name' },
  { key: (row) => row?.settings?.owner || '', label: 'Owner' },
  { key: (row) => row?.settings?.status || '', label: 'Status' },
  { key: (row) => row?.settings?.nbPlayers || '', label: 'Nb. Players' },
  { key: (row) => row?.settings?.dropTime, label: 'Drop Time' },
];

export const AuthRoom: FC<AuthRoomProps> = ({ username, onValidated }) => {
  const theme = useMantineTheme();
  const { classes } = useStyles();

  const form = useForm<AuthRoomFormValues>({
    initialValues: {
      room: '',
      pass: '',
    },
    validate: {
      room: (val) => (val.length <= 6 ? 'Room name should include at least 6 characters' : null),
    },
  });

  const [checked, setChecked] = useState(false);
  const [search, setSearch] = useState('');

  return (
    <Paper radius="md" p="md">
      <form
        onSubmit={form.onSubmit((values) => {
          onValidated(values);
        })}
      >
        <Stack>
          <Text size="lg">
            Hi{' '}
            <Text span fw={500}>
              {username}
            </Text>
            , choose or create a room to get started
          </Text>

          <Group grow position="center">
            <Stack px={20}>
              <TextInput
                placeholder="My room"
                mb="md"
                label="Room"
                {...form.getInputProps('room')}
                value={search}
                onChange={(event) => {
                  form.setFieldValue('room', event.currentTarget.value);
                  setSearch(event.currentTarget.value);
                }}
                radius="lg"
              />

              <Group spacing="xl" position="apart" className={classes.item} grow>
                <TextInput
                  placeholder="Search by any field"
                  mb="md"
                  label="Password"
                  {...form.getInputProps('password')}
                  value={form.values.pass}
                  size="xs"
                  disabled={!checked}
                  radius="lg"
                />
                <Switch
                  checked={checked}
                  onChange={(event) => setChecked(event.currentTarget.checked)}
                  color="teal"
                  size="md"
                  className={classes.switch}
                  thumbIcon={
                    checked ? (
                      <IconCheck size="0.8rem" color={theme.colors.teal[theme.fn.primaryShade()]} stroke={3} />
                    ) : (
                      <IconX size="0.8rem" color={theme.colors.red[theme.fn.primaryShade()]} stroke={3} />
                    )
                  }
                />
              </Group>

              <Stack spacing="xl" mt="xl">
                <Title fw={600} size="xl">
                  Rooms
                </Title>
                <TableSort data={roomsListData} search={form.values.room} columns={columns} />
              </Stack>
            </Stack>
          </Group>

          <Group position="center" m="md">
            <Button type="submit" size="md" radius="xl" variant="light" uppercase>
              Login
            </Button>
          </Group>
        </Stack>
      </form>
    </Paper>
  );
};
