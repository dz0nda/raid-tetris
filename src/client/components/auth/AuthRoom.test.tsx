import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { AuthRoom } from './AuthRoom';

describe('<AuthRoom /> snapshot', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(<AuthRoom username="TestUser" onValidated={jest.fn()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('<AuthRoom />', () => {
  const mockOnValidated = jest.fn();

  it('renders correctly', () => {
    const { getByText } = render(<AuthRoom username="TestUser" onValidated={mockOnValidated} />);
    expect(getByText('Hi TestUser, choose or create a room to get started')).toBeInTheDocument();
  });

  it('calls onValidated with correct values on form submission', () => {
    const { getByPlaceholderText, getByText } = render(<AuthRoom username="TestUser" onValidated={mockOnValidated} />);

    const roomInput = getByPlaceholderText('My room');
    fireEvent.change(roomInput, { target: { value: 'TestRoom' } });

    const passwordInput = getByPlaceholderText('Search by any field');
    fireEvent.change(passwordInput, { target: { value: 'TestPassword' } });

    const loginButton = getByText('Login');
    fireEvent.click(loginButton);

    expect(mockOnValidated).toHaveBeenCalledWith({ room: 'TestRoom', password: 'TestPassword' });
  });

  it('toggles the switch correctly', () => {
    const { getByRole } = render(<AuthRoom username="TestUser" onValidated={mockOnValidated} />);
    const switchButton = getByRole('checkbox');
    fireEvent.click(switchButton);
    expect(switchButton).toBeChecked();
  });

  it('updates input values correctly', () => {
    const { getByPlaceholderText } = render(<AuthRoom username="TestUser" onValidated={mockOnValidated} />);

    const roomInput = getByPlaceholderText('My room');
    fireEvent.change(roomInput, { target: { value: 'TestRoom' } });
    // expect(roomInput.value).toBe('TestRoom');

    const passwordInput = getByPlaceholderText('Search by any field');
    fireEvent.change(passwordInput, { target: { value: 'TestPassword' } });
    // expect(passwordInput.value).toBe('TestPassword');
  });
});
