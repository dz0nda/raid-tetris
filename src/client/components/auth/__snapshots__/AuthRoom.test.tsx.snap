// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthRoom /> snapshot matches snapshot 1`] = `
<div
  className="mantine-Paper-root mantine-14hio24"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="mantine-Stack-root mantine-caxjnw"
    >
      <div
        className="mantine-Text-root mantine-vgr5eq"
      >
        Hi
         
        <span
          className="mantine-Text-root mantine-1w85feo"
        >
          TestUser
        </span>
        , choose or create a room to get started
      </div>
      <div
        className="mantine-Group-root mantine-1uvkm47"
      >
        <div
          className="mantine-Stack-root mantine-1lp970e"
        >
          <div
            className="mantine-InputWrapper-root mantine-TextInput-root mantine-1erl5jz"
          >
            <label
              className="mantine-InputWrapper-label mantine-TextInput-label mantine-1fzet7j"
              htmlFor="mantine-j8j0cg6bk"
              id="mantine-j8j0cg6bk-label"
              onMouseDown={[Function]}
            >
              Room
            </label>
            <div
              className="mantine-Input-wrapper mantine-TextInput-wrapper mantine-7c7vou"
            >
              <input
                aria-invalid={false}
                className="mantine-Input-input mantine-TextInput-input mantine-e237sj"
                id="mantine-j8j0cg6bk"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="My room"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="mantine-Group-root mantine-1j8fxwl"
          >
            <div
              className="mantine-InputWrapper-root mantine-TextInput-root mantine-1erl5jz"
            >
              <label
                className="mantine-InputWrapper-label mantine-TextInput-label mantine-jkwmhw"
                htmlFor="mantine-c7doek2h1"
                id="mantine-c7doek2h1-label"
                onMouseDown={[Function]}
              >
                Password
              </label>
              <div
                className="mantine-Input-wrapper mantine-TextInput-wrapper mantine-7c7vou"
              >
                <input
                  aria-invalid={false}
                  className="mantine-Input-input mantine-TextInput-input mantine-1mzjuus"
                  data-disabled={true}
                  disabled={true}
                  id="mantine-c7doek2h1"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="Search by any field"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              className="mantine-Switch-root mantine-Switch-root mantine-1n3c3ai"
            >
              <div
                className="mantine-14n9mxy mantine-Switch-body"
              >
                <input
                  checked={false}
                  className="mantine-19htmtn mantine-Switch-input"
                  id="mantine-gjj38jxuz"
                  onChange={[Function]}
                  type="checkbox"
                />
                <label
                  className="mantine-1xnvbd7 mantine-Switch-track"
                  htmlFor="mantine-gjj38jxuz"
                >
                  <div
                    className="mantine-htyghh mantine-Switch-thumb"
                  >
                    <svg
                      className="tabler-icon tabler-icon-x"
                      fill="none"
                      height="0.8rem"
                      stroke="#fa5252"
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth={3}
                      viewBox="0 0 24 24"
                      width="0.8rem"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M18 6l-12 12"
                      />
                      <path
                        d="M6 6l12 12"
                      />
                    </svg>
                  </div>
                  <div
                    className="mantine-oxtfyg mantine-Switch-trackLabel"
                  />
                </label>
                <div
                  className="mantine-19yx22c mantine-Switch-labelWrapper"
                />
              </div>
            </div>
          </div>
          <div
            className="mantine-Stack-root mantine-62sv8f"
          >
            <h1
              className="mantine-Text-root mantine-Title-root mantine-smsr0k"
            >
              Rooms
            </h1>
            <div
              className="mantine-ScrollArea-root mantine-1deqy7c"
              dir="ltr"
              style={
                {
                  "--radix-scroll-area-corner-height": "0px",
                  "--radix-scroll-area-corner-width": "0px",
                  "position": "relative",
                }
              }
            >
              <style
                dangerouslySetInnerHTML={
                  {
                    "__html": "[data-radix-scroll-area-viewport]{scrollbar-width:none;-ms-overflow-style:none;-webkit-overflow-scrolling:touch;}[data-radix-scroll-area-viewport]::-webkit-scrollbar{display:none}",
                  }
                }
              />
              <div
                className="mantine-b6zkvl mantine-ScrollArea-viewport"
                data-radix-scroll-area-viewport=""
                style={
                  {
                    "overflowX": "scroll",
                    "overflowY": "scroll",
                  }
                }
              >
                <div
                  style={
                    {
                      "display": "table",
                      "minWidth": "100%",
                    }
                  }
                >
                  <table
                    className="mantine-Table-root mantine-1mv0s2g"
                  >
                    <thead>
                      <tr>
                        <th
                          className="mantine-1y22qhp"
                        >
                          <button
                            className="mantine-UnstyledButton-root mantine-12qguul"
                            onClick={[Function]}
                            type="button"
                          >
                            <div
                              className="mantine-Group-root mantine-1emc9ft"
                            >
                              <div
                                className="mantine-Text-root mantine-9gu3tx"
                              >
                                Name
                              </div>
                              <div
                                className="mantine-Center-root mantine-z86dzc"
                              >
                                <svg
                                  className="tabler-icon tabler-icon-selector"
                                  fill="none"
                                  height="0.9rem"
                                  stroke="currentColor"
                                  strokeLinecap="round"
                                  strokeLinejoin="round"
                                  strokeWidth={1.5}
                                  viewBox="0 0 24 24"
                                  width="0.9rem"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M8 9l4 -4l4 4"
                                  />
                                  <path
                                    d="M16 15l-4 4l-4 -4"
                                  />
                                </svg>
                              </div>
                            </div>
                          </button>
                        </th>
                        <th
                          className="mantine-1y22qhp"
                        >
                          <button
                            className="mantine-UnstyledButton-root mantine-12qguul"
                            onClick={[Function]}
                            type="button"
                          >
                            <div
                              className="mantine-Group-root mantine-1emc9ft"
                            >
                              <div
                                className="mantine-Text-root mantine-9gu3tx"
                              >
                                Owner
                              </div>
                              <div
                                className="mantine-Center-root mantine-z86dzc"
                              >
                                <svg
                                  className="tabler-icon tabler-icon-selector"
                                  fill="none"
                                  height="0.9rem"
                                  stroke="currentColor"
                                  strokeLinecap="round"
                                  strokeLinejoin="round"
                                  strokeWidth={1.5}
                                  viewBox="0 0 24 24"
                                  width="0.9rem"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M8 9l4 -4l4 4"
                                  />
                                  <path
                                    d="M16 15l-4 4l-4 -4"
                                  />
                                </svg>
                              </div>
                            </div>
                          </button>
                        </th>
                        <th
                          className="mantine-1y22qhp"
                        >
                          <button
                            className="mantine-UnstyledButton-root mantine-12qguul"
                            onClick={[Function]}
                            type="button"
                          >
                            <div
                              className="mantine-Group-root mantine-1emc9ft"
                            >
                              <div
                                className="mantine-Text-root mantine-9gu3tx"
                              >
                                Status
                              </div>
                              <div
                                className="mantine-Center-root mantine-z86dzc"
                              >
                                <svg
                                  className="tabler-icon tabler-icon-selector"
                                  fill="none"
                                  height="0.9rem"
                                  stroke="currentColor"
                                  strokeLinecap="round"
                                  strokeLinejoin="round"
                                  strokeWidth={1.5}
                                  viewBox="0 0 24 24"
                                  width="0.9rem"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M8 9l4 -4l4 4"
                                  />
                                  <path
                                    d="M16 15l-4 4l-4 -4"
                                  />
                                </svg>
                              </div>
                            </div>
                          </button>
                        </th>
                        <th
                          className="mantine-1y22qhp"
                        >
                          <button
                            className="mantine-UnstyledButton-root mantine-12qguul"
                            onClick={[Function]}
                            type="button"
                          >
                            <div
                              className="mantine-Group-root mantine-1emc9ft"
                            >
                              <div
                                className="mantine-Text-root mantine-9gu3tx"
                              >
                                Nb. Players
                              </div>
                              <div
                                className="mantine-Center-root mantine-z86dzc"
                              >
                                <svg
                                  className="tabler-icon tabler-icon-selector"
                                  fill="none"
                                  height="0.9rem"
                                  stroke="currentColor"
                                  strokeLinecap="round"
                                  strokeLinejoin="round"
                                  strokeWidth={1.5}
                                  viewBox="0 0 24 24"
                                  width="0.9rem"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M8 9l4 -4l4 4"
                                  />
                                  <path
                                    d="M16 15l-4 4l-4 -4"
                                  />
                                </svg>
                              </div>
                            </div>
                          </button>
                        </th>
                        <th
                          className="mantine-1y22qhp"
                        >
                          <button
                            className="mantine-UnstyledButton-root mantine-12qguul"
                            onClick={[Function]}
                            type="button"
                          >
                            <div
                              className="mantine-Group-root mantine-1emc9ft"
                            >
                              <div
                                className="mantine-Text-root mantine-9gu3tx"
                              >
                                Drop Time
                              </div>
                              <div
                                className="mantine-Center-root mantine-z86dzc"
                              >
                                <svg
                                  className="tabler-icon tabler-icon-selector"
                                  fill="none"
                                  height="0.9rem"
                                  stroke="currentColor"
                                  strokeLinecap="round"
                                  strokeLinejoin="round"
                                  strokeWidth={1.5}
                                  viewBox="0 0 24 24"
                                  width="0.9rem"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M8 9l4 -4l4 4"
                                  />
                                  <path
                                    d="M16 15l-4 4l-4 -4"
                                  />
                                </svg>
                              </div>
                            </div>
                          </button>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr>
                        <td>
                          dzsazs
                        </td>
                        <td>
                          Alice
                        </td>
                        <td>
                          active
                        </td>
                        <td>
                          4
                        </td>
                        <td>
                          5000
                        </td>
                      </tr>
                      <tr>
                        <td>
                          fdfsdf
                        </td>
                        <td>
                          Bob
                        </td>
                        <td>
                          waiting
                        </td>
                        <td>
                          2
                        </td>
                        <td>
                          6000
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="mantine-Group-root mantine-gnwlu3"
      >
        <button
          className="mantine-UnstyledButton-root mantine-Button-root mantine-j4t094"
          data-button={true}
          type="submit"
        >
          <div
            className="mantine-1wpc1xj mantine-Button-inner"
          >
            <span
              className="mantine-1ryt1ht mantine-Button-label"
              style={
                {
                  "textTransform": "uppercase",
                }
              }
            >
              Login
            </span>
          </div>
        </button>
      </div>
    </div>
  </form>
</div>
`;
